### PR TITLE
Image* refactoring, ImageRepository tag history

### PR DESCRIPTION
**Pull request by ncdc**
Rework image rest/storage code to use new upstream patterns.

Store tag history in ImageRepository.Status.

Add DockerImageManifest to DockerImage to prepare for v2 registry
integration.

For https://trello.com/c/ZPKh3cBp/296-3-management-of-historical-image-repository-tags-imageregistry-beta2

### CODE REVIEW

**Comment by ncdc** [comment]
@smarterclayton PTAL. I still need to do the virtual pretty image we discussed, although maybe that could go in a follow-on PR.

**Comment by ncdc** [comment]
@smarterclayton latest commit returns a prettified DockerImageMetadata for ImageRepositoryTags and ImageStreamImages; Images returns the raw manifest json

**Comment by smarterclayton** [comment]
This is fine otherwise, squash where necessary, address any cleanup, and let me know and I'll merge.

**Comment by ncdc** [comment]
I'm working on adding some test cases. I'll finish them and squash tomorrow.

File: pkg/image/api/helper.go [hunk 1]
```diff
@@ -69,23 +70,43 @@ func JoinDockerPullSpec(registry, namespace, name, tag string) string {
...
-	if len(parts) == 2 {
-		image = parts[0]
-		tag = parts[1]
+
+	manifestData := image.DockerImageManifest
+
+	image.DockerImageManifest = ""
+
+	manifest := DockerImageManifest{}
+	if err := json.Unmarshal([]byte(image.DockerImageManifest), &manifest); err != nil {
```
Comment (mfojtik) [inline pkg/image/api/helper.go]: If you set the `image.DockerImageManifest` to "" on 80, isn't this Unmarshal always unmarshaling the empty string?

File: pkg/image/api/types.go [hunk 2]
```diff
@@ -54,6 +57,23 @@ type ImageRepositoryStatus struct {
 	// Represents the effective location this repository may be accessed at. May be empty until the server
 	// determines where the repository is located
 	DockerImageRepository string `json:"dockerImageRepository,omitempty"`
+	// A historical record of images associated with each tag. The first entry in the TagEvent array is
+	// the currently tagged image.
+	Tags map[string]TagEventList `json:"tags,omitempty"`
+}
+
+type TagEventList struct {
```
Comment (mfojtik) [inline pkg/image/api/types.go]: Add godoc for this?

File: pkg/image/registry/imagerepository/strategy.go [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+		repo.Status.Tags = make(map[string]api.TagEventList)
+	}
+
+	repo.Status.DockerImageRepository = s.dockerImageRepository(repo)
+	updateTagHistory(repo)
+
+	return validation.ValidateImageRepositoryUpdate(repo, oldRepo)
+}
+
+func (s imageRepositoryStrategy) Decorate(obj runtime.Object) error {
```
Comment (mfojtik) [inline pkg/image/registry/imagerepository/strategy.go]: godoc

File: pkg/image/api/helper.go (hunk 1 above)
Comment (ncdc) [inline pkg/image/api/helper.go]: Yes, I haven't finished refactoring this function.

File: pkg/image/api/types.go (hunk 2 above)
Comment (ncdc) [inline pkg/image/api/types.go]: Don't worry, there will be godoc :-)

**Comment by ncdc** [comment]
[test]

**Comment by ncdc** [comment]
@smarterclayton updated, PTAL

**Comment by ncdc** [comment]
Just realized I need to update the client code to not use Images with a namespacer... working on that.

**Comment by ncdc** [comment]
Removed references to images with a namespace.

**Comment by ncdc** [comment]
@smarterclayton ready for review. Note, I still need to make the auth changes so images are locked down and users are required to use imageStreamImages and imageRepositoryTags. I'll do that in a follow-on PR.

File: pkg/image/api/v1beta1/types.go [hunk 4]
```diff
@@ -55,9 +58,26 @@ type ImageRepositoryStatus struct {
...
+}
+
+// TagEvent is used by ImageRepositoryStatus to keep a historical record of images associated with a tag.
+type TagEvent struct {
+	// When the TagEvent was created
+	Created util.Time `json:"created"`
+	// The string that can be used to pull this image
+	DockerImageReference string `json:"dockerImageReference"`
+	// The image ID
+	DockerImageID string `json:"dockerImageID"`
```
Comment (smarterclayton) [inline pkg/image/api/v1beta1/types.go]: Not a docker image id.  An openshift image.  So "Image"

**Comment by smarterclayton** [comment]
LGTM [merge]

**Pull request merged**